### PR TITLE
Merge H3-index branch to master

### PR DESCRIPTION
**Pull request by yupeng9**
Use this PR instead of the previous [PR](https://github.com/apache/incubator-pinot/pull/6376) because the last commit mistakenly merged master (instead of rebase).

## Description
Adding support for indexing (lat,lon) using Uber's H3 indexing. Part of https://github.com/apache/incubator-pinot/issues/5280

```
select count(*) from starbucksStores  where ST_DISTANCE(location_st_point , ST_POINT(-122, 37)) < 10
```

Check the following individual PRs for more details
https://github.com/apache/incubator-pinot/pull/6387
https://github.com/apache/incubator-pinot/pull/6391
https://github.com/apache/incubator-pinot/pull/6397
https://github.com/apache/incubator-pinot/pull/6398
https://github.com/apache/incubator-pinot/pull/6400

## Upgrade Notes
- New configuration options for Geospatial indexing using H3
* [X] Yes 

## Release Notes
If you have tagged this as either backward-incompat or release-notes,
Add an [H3](https://h3geo.org/)-based geospatial indexing to speed up the filtering with `ST_Distance`

### CODE REVIEW

**Comment by yupeng9** [comment]
@kishoreg Comments addressed. PTAL

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/datasource/BaseDataSource.java [hunk 1]
```diff
@@ -46,19 +48,29 @@ public BaseDataSource(DataSourceMetadata dataSourceMetadata, ForwardIndexReader<
       @Nullable Dictionary dictionary, @Nullable InvertedIndexReader<?> invertedIndex,
       @Nullable InvertedIndexReader<?> rangeIndex, @Nullable TextIndexReader textIndex,
       @Nullable TextIndexReader fstIndex, @Nullable JsonIndexReader jsonIndex, @Nullable BloomFilterReader bloomFilter,
-      @Nullable NullValueVectorReader nullValueVector) {
+      @Nullable NullValueVectorReader nullValueVector, @Nullable H3IndexReader h3Index) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/datasource/BaseDataSource.java:51]: Put argument `h3Index` between `rangeIndex` and `textIndex`

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/datasource/BaseDataSource.java [hunk 2]
```diff
@@ -46,19 +48,29 @@ public BaseDataSource(DataSourceMetadata dataSourceMetadata, ForwardIndexReader<
...
     _rangeIndex = rangeIndex;
+    _h3Index = h3Index;
     _textIndex = textIndex;
     _fstIndex = fstIndex;
     _jsonIndex = jsonIndex;
     _bloomFilter = bloomFilter;
     _nullValueVector = nullValueVector;
   }
 
+  public BaseDataSource(DataSourceMetadata dataSourceMetadata, ForwardIndexReader<?> forwardIndex,
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/datasource/BaseDataSource.java:65]: Remove this constructor. We cannot have a separate constructor per nullable index. We can consider using builder, but out of the scope of this PR

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/datasource/BaseDataSource.java [hunk 3]
```diff
@@ -105,6 +117,10 @@ public JsonIndexReader getJsonIndex() {
     return _jsonIndex;
   }
 
+  public H3IndexReader getH3Index() {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/datasource/BaseDataSource.java:120]: Put this before text index for consistency

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/SegmentPreProcessor.java [hunk 4]
```diff
@@ -115,7 +116,13 @@ public void process()
           new RangeIndexHandler(_indexDir, _segmentMetadata, _indexLoadingConfig, segmentWriter);
       rangeIndexHandler.createRangeIndices();
 
-      // Create text indices according to the index config.
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/SegmentPreProcessor.java:118]: Keep the original comment for text index?

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/FieldConfig.java [hunk 5]
```diff
@@ -67,7 +67,7 @@ public FieldConfig(@JsonProperty(value = "name", required = true) String name,
 
   // If null, there won't be any index
   public enum IndexType {
-    INVERTED, SORTED, TEXT, FST
+    INVERTED, SORTED, TEXT, FST, H3
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/FieldConfig.java:70]: High level question, is it possible to have more than one of these index types on the same column?

**Review by Jackie-Jiang — COMMENTED** [review]
Partially reviewed

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java [hunk 6]
```diff
@@ -1187,6 +1189,8 @@ public LLRealtimeSegmentDataManager(RealtimeSegmentZKMetadata segmentZKMetadata,
     Set<String> textIndexColumns = indexLoadingConfig.getTextIndexColumns();
     _textIndexColumns = new ArrayList<>(textIndexColumns);
 
+    _h3IndexColumns = indexLoadingConfig.getH3IndexColumns();
```
[the remark this answers is not in the document]
Comment (yupeng9) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java:1192]: I know. Keep it for the same style.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/datasource/BaseDataSource.java (hunk 1 above)
Comment (yupeng9) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/datasource/BaseDataSource.java:51]: this is actually intentional to use append. It's easier for subclass to determine the differences.

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java:1192]: Not really the same style, the other member variables are needed for `RealtimeSegmentConverter`. Also don't put it between text and fst as they are both for text index

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/datasource/BaseDataSource.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/datasource/BaseDataSource.java:51]: Since all the subclasses are internal, we'd better keep them the same order for easier maintenance in the future as we add more indexes.

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (yupeng9) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java:1192]: I see. That makes sense.

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/RealtimeSegmentConfig.java [hunk 7]
```diff
@@ -63,7 +68,7 @@ private RealtimeSegmentConfig(String tableNameWithType, String segmentName, Stri
       boolean offHeap, PinotDataBufferMemoryManager memoryManager, RealtimeSegmentStatsHistory statsHistory,
       String partitionColumn, PartitionFunction partitionFunction, int partitionId, boolean aggregateMetrics,
       boolean nullHandlingEnabled, String consumerDir, UpsertConfig.Mode upsertMode,
-      PartitionUpsertMetadataManager partitionUpsertMetadataManager) {
+      PartitionUpsertMetadataManager partitionUpsertMetadataManager, List<H3IndexColumn> h3IndexColumns) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/RealtimeSegmentConfig.java:71]: Move the argument after `jsonIndexColumns`

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/RealtimeSegmentConfig.java [hunk 8]
```diff
@@ -359,12 +375,17 @@ public Builder setPartitionUpsertMetadataManager(PartitionUpsertMetadataManager
       return this;
     }
 
+    public Builder setH3IndexColumns(List<H3IndexColumn> h3IndexColumns) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/RealtimeSegmentConfig.java:378]: Move it after `setJsonIndexColumns`

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/geospatial/RealtimeH3IndexReader.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    _lowestResolution = resolution.getLowestResolution();
+    _h3Core = H3Core.newInstance();
+  }
+
+  @Override
+  public void add(int docId, double lat, double lon) {
+    // TODO support multiple resolutions
+    Long h3Id = _h3Core.geoToH3(lat, lon, _lowestResolution);
+    _h3IndexMap.computeIfAbsent(h3Id, k -> new ThreadSafeMutableRoaringBitmap());
+    synchronized (this) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/geospatial/RealtimeH3IndexReader.java:56]: No need to synchronize it. It is guarded by `ConcurrentHashMap` and `ThreadSafeMutableRoaringBitmap`

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/geospatial/RealtimeH3IndexReader.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * A H3 index reader for the real-time H3 index values on the fly.
+ * <p>This class is thread-safe for single writer multiple readers.
+ */
+public class RealtimeH3IndexReader implements GeoSpatialIndexCreator, H3IndexReader {
+  private final H3Core _h3Core;
+  private final Map<Long, ThreadSafeMutableRoaringBitmap> _h3IndexMap = new ConcurrentHashMap<>();
+  private final H3IndexResolution _resolution;
+  private int _lowestResolution;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/geospatial/RealtimeH3IndexReader.java:42]: final

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/GeoSpatialIndexCreator.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+ */
+public interface GeoSpatialIndexCreator extends Closeable {
+
+  /**
+   * Adds a doc associated with a geospatial point
+   * @param docId the document id
+   * @param lat the latitude
+   * @param lon the longitude
+   */
+  void add(int docId, double lat, double lon);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/creator/GeoSpatialIndexCreator.java:35]: We don't usually put docId explicitly. It is always appended as the next doc

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    _segment = indexSegment;
+    FunctionContext function = predicate.getLhs().getFunction();
+    String columnName;
+
+    // TODO: handle composite function that contains ST_DISTANCE
+    if (function.getArguments().get(0).getType() == ExpressionContext.Type.IDENTIFIER) {
+      columnName = function.getArguments().get(0).getIdentifier();
+      byte[] bytes = BytesUtils.toBytes(function.getArguments().get(1).getLiteral());
+      _geometry = GeometrySerializer.deserialize(bytes);
+    } else if (function.getArguments().get(1).getType() == ExpressionContext.Type.IDENTIFIER) {
```
[the remark this answers is not in the document]
Comment (yupeng9) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java:73]: Right, `FilterPlanNode.canApplyH3Index ` is more restrictive, so I don't do the same checking here.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/GeoSpatialIndexCreator.java (hunk 11 above)
Comment (yupeng9) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/creator/GeoSpatialIndexCreator.java:35]: But how does that work? The index creator keeps the mapping of `H3Index -> BitMap of docIds`

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/geospatial/MutableH3Index.java [hunk 13]
```diff
@@ -18,48 +18,53 @@
...
 
 
 /**
  * A H3 index reader for the real-time H3 index values on the fly.
  * <p>This class is thread-safe for single writer multiple readers.
  */
-public class RealtimeH3IndexReader implements GeoSpatialIndexCreator, H3IndexReader {
-  private final H3Core _h3Core;
-  private final Map<Long, ThreadSafeMutableRoaringBitmap> _h3IndexMap = new ConcurrentHashMap<>();
+public class MutableH3Index implements H3IndexReader {
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/geospatial/MutableH3Index.java:39]: shall this implement GeoSpatialIndexCreator?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java [hunk 14]
```diff
@@ -260,6 +274,19 @@ public void init(SegmentGeneratorConfig segmentCreationSpec, SegmentIndexCreatio
...
 
+      H3IndexConfig h3IndexConfig = h3IndexConfigs.get(columnName);
+      if (h3IndexConfig != null) {
+        Preconditions
+            .checkState(fieldSpec.isSingleValueField(), "H3 index is currently only supported on single-value columns");
+        Preconditions.checkState(fieldSpec.getDataType() == DataType.BYTES,
+            "H3 index is currently only supported on BYTES columns");
+        H3IndexResolution resolution = h3IndexConfig.getResolution();
+        GeoSpatialIndexCreator h3IndexCreator =
+            segmentCreationSpec.isOnHeap() ? new OnHeapH3IndexCreator(_indexDir, columnName, resolution)
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java:285]: any scenario to use on heap index?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/H3IndexHandler.java [hunk 15]
```diff
@@ -86,72 +88,77 @@ public void createH3Indices()
...
       }
 
       // Create a marker file.
       FileUtils.touch(inProgress);
     } else {
       // Marker file exists, which means last run gets interrupted.
-      // Remove range index if exists.
+
+      // Remove H3 index if exists.
       // For v1 and v2, it's the actual range index. For v3, it's the temporary range index.
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/H3IndexHandler.java:110]: remote this line?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/BitmapInvertedIndexReader.java [hunk 16]
```diff
@@ -18,84 +18,71 @@
...
+  public BitmapInvertedIndexReader(PinotDataBuffer dataBuffer, int numBitmaps) {
+    _numBitmaps = numBitmaps;
 
-    int lastOffset = _dataBuffer.getInt(_numBitmaps * Integer.BYTES);
-    Preconditions.checkState(lastOffset == _dataBuffer.size(),
-        "The last offset should be equal to buffer size! Current lastOffset: " + lastOffset + ", buffer size: "
-            + _dataBuffer.size());
+    long offsetBufferEndOffset = (long) (numBitmaps + 1) * Integer.BYTES;
+    _offsetBuffer = dataBuffer.view(0, offsetBufferEndOffset, ByteOrder.BIG_ENDIAN);
+    _bitmapBuffer = dataBuffer.view(offsetBufferEndOffset, dataBuffer.size());
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/BitmapInvertedIndexReader.java:52]: are these changes related to this PR?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java [hunk 17]
```diff
@@ -50,94 +46,164 @@
...
-        fullMatchedDocIds.or(docIds);
+      return new FilterBlock(EmptyDocIdSet.getInstance());
+    }
+
+    if (Double.isNaN(_lowerBound) || _lowerBound < 0) {
+      // No lower bound
+
+      if (Double.isNaN(_upperBound)) {
+        // No upper bound
+        return new FilterBlock(new MatchAllDocIdSet(_numDocs));
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java:103]: shall we have some tests for this algorithm?

File: pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/MutableSegmentImpl.java [hunk 18]
```diff
@@ -47,15 +48,15 @@
...
+import org.apache.pinot.core.realtime.impl.geospatial.MutableH3Index;
 import org.apache.pinot.core.realtime.impl.invertedindex.RealtimeInvertedIndexReader;
 import org.apache.pinot.core.realtime.impl.invertedindex.RealtimeLuceneIndexRefreshState;
 import org.apache.pinot.core.realtime.impl.invertedindex.RealtimeLuceneIndexRefreshState.RealtimeLuceneReaders;
 import org.apache.pinot.core.realtime.impl.invertedindex.RealtimeLuceneTextIndexReader;
 import org.apache.pinot.core.realtime.impl.json.MutableJsonIndex;
 import org.apache.pinot.core.realtime.impl.nullvalue.MutableNullValueVector;
 import org.apache.pinot.core.segment.creator.impl.V1Constants;
-import org.apache.pinot.core.segment.creator.impl.geospatial.H3IndexResolution;
+import org.apache.pinot.core.segment.creator.impl.inv.geospatial.H3IndexConfig;
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/MutableSegmentImpl.java:59]: One type of inverted index (from value to docIds)

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/BitmapInvertedIndexReader.java (hunk 16 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/BitmapInvertedIndexReader.java:52]: Yes, to be able to read the bitmap index within the H3 index

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/H3IndexHandler.java (hunk 15 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/H3IndexHandler.java:110]: Done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java (hunk 14 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java:285]: When memory is not an issue (e.g. hadoop), on-heap creator is more efficient

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/geospatial/MutableH3Index.java (hunk 13 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/geospatial/MutableH3Index.java:39]: We don't mix mutable index with index creator. Index creator is for offline segment creation, and have a `seal()` method

File: pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java [hunk 19]
```diff
@@ -93,27 +93,37 @@ public BaseFilterOperator run() {
     }
   }
 
+  /**
+   * H3 index can be applied iff:
+   * <ul>
+   *   <li>Predicate is of type RANGE</li>
+   *   <li>Left-hand-side of the predicate is an ST_Distance function</li>
+   *   <li>One argument of the ST_Distance function is an identifier, the other argument is an literal</li>
+   *   <li>The identifier column has H3 index</li>
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java:102]: There is no easy way to support it because the index has to be generated on a column, but not on a function

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java [hunk 20]
```diff
@@ -50,94 +46,164 @@
...
 
-  public H3IndexFilterOperator(Predicate predicate, IndexSegment indexSegment, int numDocs) {
+  public H3IndexFilterOperator(IndexSegment segment, Predicate predicate, int numDocs) {
+    _segment = segment;
     _predicate = predicate;
-    _segment = indexSegment;
-    FunctionContext function = predicate.getLhs().getFunction();
-    String columnName;
-
-    // TODO: handle composite function that contains ST_DISTANCE
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java]: Added

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java [hunk 21]
```diff
@@ -50,94 +46,164 @@
...
-    List<Long> fullMatchRings = new ArrayList<>();
-    if (numFullMatchedRings > 0) {
-      fullMatchRings = _h3Core.kRing(h3Id, numFullMatchedRings - 1);
-      for (long id : fullMatchRings) {
-        ImmutableRoaringBitmap docIds = _h3IndexReader.getDocIds(id);
-        fullMatchedDocIds.or(docIds);
+      return new FilterBlock(EmptyDocIdSet.getInstance());
+    }
+
+    if (Double.isNaN(_lowerBound) || _lowerBound < 0) {
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java:98]: We need to handle the case for distance > 0, and not return the same point

File: pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (yupeng9) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java:102]: oh, i meant `ST_DISTANCE(toSphericalGeography(location)`, in which `location` is a geometry

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java [hunk 22]
```diff
@@ -50,94 +46,164 @@
...
-      // Override this method to reflect the entries scanned
+    return getFilterBlock(fullMatchDocIds, partialMatchDocIds);
+  }
+
+  /**
+   * Returns the H3 ids that is ALWAYS fully covered by the circle with the given distance as the radius and a point
+   * within the _h3Id hexagon as the center.
+   */
+  private List<Long> getFullMatchH3Ids(double distance) {
+    // NOTE: Pick a constant slightly larger than sqrt(3) to be conservative
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java:178]: Added

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java [hunk 23]
```diff
@@ -50,94 +46,164 @@
...
+    }
+  }
+
+  /**
+   * Returns the H3 ids that MIGHT BE partially/fully covered by the circle with the given distance as the radius and a
+   * point within the _h3Id hexagon as the center.
+   */
+  private List<Long> getPartialMatchH3Ids(double distance) {
+    // NOTE: Add a small delta (0.001) to be conservative
+    int numRings = (int) Math.floor((distance / _edgeLength + 2) / 1.5 + 0.001);
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java:193]: Added

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java (hunk 17 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java:103]: Added

File: pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java:102]: I see what you mean. There is a TODO on line 116 for that.
IMO, H3 index can only be applied to geography, but not geometry. Not sure if we need geometry at all.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java [hunk 24]
```diff
@@ -90,110 +92,139 @@ public H3IndexFilterOperator(IndexSegment segment, Predicate predicate, int numD
...
+  }
+
+  /**
+   * Returns the H3 ids for the given number of rings around the _h3Id.
+   * IMPORTANT: Throw exception when number of rings is too large because H3 library might send SIGILL for large number
+   * of rings, which can terminate the JVM. Also, the result won't be accurate when the distance is too large. In such
+   * case, the operator will fallback to ExpressionFilterOperator.
+   */
+  private List<Long> getH3Ids(int numRings) {
+    Preconditions.checkState(numRings <= 100, "Expect numRings <= 100, got: %s", numRings);
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java:221]: Will this fall back to scan-based approach?

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java:221]: Yes, as mentioned in the comments

**Comment by yupeng9** [comment]
LGTM

**Review by Jackie-Jiang — APPROVED** [review]
